src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -191,145 +191,71 @@ public ICompilationTracker FreezePartialState(CancellationToken cancellationToke
-                    state,
-                    out var inProgressProject,
-                    out var compilationWithoutGeneratedDocuments,
-                    out var compilationWithGeneratedDocuments,
-                    out var generatorInfo,
-                    cancellationToken);
-
-                ImmutableList<(ProjectState oldState, CompilationAndGeneratorDriverTranslationAction action)> pendingActions;
-                if (inProgressProject.DocumentStates.TryGetState(docState.Id, out oldState))
-                {
-                    // The document had been previously parsed and it's there, so we can update it with our current
-                    // state. Note if no compilation existed GetPartialCompilationState would have produced an empty
-                    // one, and removed any documents, so inProgressProject.DocumentStates would have been empty
-                    // originally.
-                    pendingActions = [(oldState: inProgressProject, new CompilationAndGeneratorDriverTranslationAction.TouchDocumentAction(oldState, docState))];
-                    inProgressProject = inProgressProject.UpdateDocument(docState, contentChanged: true);
-                }
-                else
+                var projectState = state switch
                 {
-                    // The document wasn't present in the original snapshot at all, and we just need to add the
-                    // document.
-                    pendingActions = [(oldState: inProgressProject, new CompilationAndGeneratorDriverTranslationAction.AddDocumentsAction([docState]))];
-                    inProgressProject = inProgressProject.AddDocuments([docState]);
-                }
-
-                // Transition us to a frozen in progress state.  With the best compilations up to this point, and the
-                // pending actions we'll need to perform on it to get to the final state.
-                var inProgressState = InProgressState.Create(
-                    isFrozen: true,
-                    compilationWithoutGeneratedDocuments,
-                    generatorInfo,
-                    compilationWithGeneratedDocuments,
-                    pendingActions);
+                    // If we don't have an existing state, then transition to a project state without any data.
+                    null => this.ProjectState.RemoveAllDocuments(),
+                    FinalCompilationTrackerState => this.ProjectState,
+                    // If we have an in progress state with no steps, then we're just at the current project state.
+                    InProgressState { PendingTranslationSteps: [] } => this.ProjectState,
+                    // Otherwise, reset us to whatever state the InProgressState had currently transitioned to.
+                    InProgressState inProgressState => inProgressState.PendingTranslationSteps.First().oldState,
+                    _ => throw ExceptionUtilities.UnexpectedValue(state.GetType()),
+                };
```

Comment:
next step, we break out which ProjectState we should transition to. Note @jasonmalinowski  i want to discuss the `InProgressState inProgressState => inProgressState.PendingTranslationSteps.First().oldState,` logic.